File: medical_logic/medical_logic/report/medical_summary_report/medical_summary_report.py
"""
medical_logic/medical_logic/report/medical_summary_report/medical_summary_report.py

Medical Summary Report — month-wise aggregated totals.
One row per month showing employee count, claim count and total claimed.
"""

import frappe
from frappe import _
from frappe.utils import flt


MONTH_ORDER = [
    "January", "February", "March",    "April",
    "May",     "June",     "July",     "August",
    "September","October", "November", "December",
]


def execute(filters=None):
    filters = filters or {}
    validate_filters(filters)
    columns = get_columns()
    data    = get_data(filters)
    summary = get_summary(data)
    return columns, data, None, None, summary


# ─────────────────────────────────────────────────────────────────────────────
# VALIDATE
# ─────────────────────────────────────────────────────────────────────────────

def validate_filters(filters):
    if not filters.get("year"):
        frappe.throw(_("Please select a Year to generate the report."))


# ─────────────────────────────────────────────────────────────────────────────
# COLUMNS
# ─────────────────────────────────────────────────────────────────────────────

def get_columns():
    return [
        {
            "label":     _("Month"),
            "fieldname": "month",
            "fieldtype": "Data",
            "width":     120,
        },
        {
            "label":     _("Year"),
            "fieldname": "year",
            "fieldtype": "Data",
            "width":     70,
        },
        {
            "label":     _("Employee ID"),
            "fieldname": "employee",
            "fieldtype": "Link",
            "options":   "Employee",
            "width":     110,
        },
        {
            "label":     _("Employee Name"),
            "fieldname": "employee_name",
            "fieldtype": "Data",
            "width":     180,
        },
        {
            "label":     _("Department"),
            "fieldname": "department",
            "fieldtype": "Data",
            "width":     150,
        },
        {
            "label":     _("Total Claims"),
            "fieldname": "claim_count",
            "fieldtype": "Int",
            "width":     110,
        },
        {
            "label":     _("Total Medical Claimed (Rs)"),
            "fieldname": "total_claimed",
            "fieldtype": "Currency",
            "width":     200,
        },
    ]


# ─────────────────────────────────────────────────────────────────────────────
# DATA
# ─────────────────────────────────────────────────────────────────────────────

def get_data(filters):
    year       = filters.get("year")
    from_month = filters.get("from_month")
    to_month   = filters.get("to_month")
    employee   = filters.get("employee")
    department = filters.get("department")

    extra = ""
    if employee:
        extra += " AND ec.employee = %(employee)s"
    if department:
        extra += " AND ec.department = %(department)s"

    sql_filters = {
        "year":       year,
        "employee":   employee,
        "department": department,
    }

    rows = frappe.db.sql(f"""
        SELECT
            ec.custom_bill_of_month                         AS month,
            %(year)s                                        AS year,
            ec.employee                                     AS employee,
            ec.employee_name                                AS employee_name,
            ec.department                                   AS department,
            COUNT(DISTINCT ec.name)                         AS claim_count,
            SUM(
                CASE
                    WHEN IFNULL(ecd.sanctioned_amount, 0) > 0
                    THEN ecd.sanctioned_amount
                    ELSE IFNULL(ecd.amount, 0)
                END
            )                                               AS total_claimed
        FROM
            `tabExpense Claim`        ec
        JOIN
            `tabExpense Claim Detail` ecd ON ecd.parent = ec.name
        WHERE
            ec.docstatus              != 2
            AND YEAR(ec.creation)      = %(year)s
            AND ec.custom_bill_of_month IS NOT NULL
            AND ec.custom_bill_of_month != ''
            {extra}
        GROUP BY
            ec.custom_bill_of_month, ec.employee
        ORDER BY
            ec.employee_name
    """, sql_filters, as_dict=True)

    # Sort by calendar month order
    month_index = {m: i for i, m in enumerate(MONTH_ORDER)}
    rows = sorted(rows, key=lambda r: month_index.get(r.get("month") or "", 99))

    # Apply from_month / to_month slice
    if from_month or to_month:
        from_idx = month_index.get(from_month, 0)   if from_month else 0
        to_idx   = month_index.get(to_month,   11)  if to_month   else 11
        rows = [
            r for r in rows
            if from_idx <= month_index.get(r.get("month") or "", 99) <= to_idx
        ]

    for r in rows:
        r["total_claimed"] = flt(r["total_claimed"], 2)

    return rows


# ─────────────────────────────────────────────────────────────────────────────
# SUMMARY
# ─────────────────────────────────────────────────────────────────────────────

def get_summary(data):
    if not data:
        return []

    total_claimed   = sum(flt(r.get("total_claimed")  or 0) for r in data)
    total_claims    = sum(int(r.get("claim_count")     or 0) for r in data)
    total_employees = sum(int(r.get("employee_count")  or 0) for r in data)
    total_months    = len(data)

    return [
        {
            "label":    _("Months"),
            "value":    total_months,
            "indicator": "blue",
            "datatype": "Int",
        },
        {
            "label":    _("Total Claims"),
            "value":    total_claims,
            "indicator": "orange",
            "datatype": "Int",
        },
        {
            "label":    _("Total Employees Claimed"),
            "value":    total_employees,
            "indicator": "purple",
            "datatype": "Int",
        },
        {
            "label":    _("Grand Total Medical Claimed"),
            "value":    total_claimed,
            "indicator": "green",
            "datatype": "Currency",
        },
    ]

File: medical_logic/medical_logic/report/medical_summary_report/medical_summary_report.js
// medical_logic/medical_logic/report/medical_summary_report/medical_summary_report.js

frappe.query_reports["Medical Summary Report"] = {

    filters: [
        {
            fieldname:  "year",
            label:      __("Year"),
            fieldtype:  "Select",
            options:    get_year_options(),
            default:    frappe.datetime.get_today().split("-")[0],
            reqd:       1,
        },
        {
            fieldname: "from_month",
            label:     __("From Month"),
            fieldtype: "Select",
            options: [
                "",
                "January","February","March","April","May","June",
                "July","August","September","October","November","December",
            ].join("\n"),
        },
        {
            fieldname: "to_month",
            label:     __("To Month"),
            fieldtype: "Select",
            options: [
                "",
                "January","February","March","April","May","June",
                "July","August","September","October","November","December",
            ].join("\n"),
        },
        {
            fieldname: "employee",
            label:     __("Employee"),
            fieldtype: "Link",
            options:   "Employee",
        },
        {
            fieldname: "department",
            label:     __("Department"),
            fieldtype: "Link",
            options:   "Department",
        },
    ],

    // ── Print Button ──────────────────────────────────────────────────────────
    onload: function(report) {
        report.page.add_inner_button(__("🖨️  Print"), function() {
            var data    = frappe.query_report.data;
            var columns = frappe.query_report.columns;

            if (!data || !data.length) {
                frappe.msgprint(__("No data to print."));
                return;
            }

            var fv = frappe.query_report.get_filter_values();
            var filters_html = Object.keys(fv)
                .filter(k => fv[k])
                .map(k => `<span style="margin-right:18px"><b>${__(k.replace(/_/g," "))}:</b> ${fv[k]}</span>`)
                .join("");

            var thead = "<tr>" + columns.map(c => `<th>${__(c.label)}</th>`).join("") + "</tr>";

            var grand_total = 0;
            var tbody = data.map(function(row) {
                grand_total += parseFloat(row["total_claimed"] || 0);
                return "<tr>" + columns.map(function(col) {
                    var val = row[col.fieldname];
                    if (val == null) val = "";
                    if (col.fieldtype === "Currency") {
                        val = "Rs " + parseFloat(val || 0).toLocaleString("en-PK", {
                            minimumFractionDigits: 2, maximumFractionDigits: 2
                        });
                    }
                    return `<td>${val}</td>`;
                }).join("") + "</tr>";
            }).join("");

            // Grand total row
            var total_row = columns.map(function(col, i) {
                if (col.fieldname === "total_claimed") {
                    return `<td style="font-weight:bold;color:#5E64FF">Rs ${grand_total.toLocaleString("en-PK",{minimumFractionDigits:2,maximumFractionDigits:2})}</td>`;
                } else if (i === 0) {
                    return `<td style="font-weight:bold">Grand Total</td>`;
                }
                return "<td></td>";
            }).join("");

            var print_html = `
            <html><head><title>Medical Summary Report</title>
            <style>
                body   { font-family:Arial,sans-serif; font-size:11px; margin:20px; }
                h2     { text-align:center; font-size:15px; margin-bottom:2px; }
                h3     { text-align:center; font-size:12px; color:#444; margin:0 0 4px; }
                .meta  { text-align:center; font-size:10px; color:#999; margin-bottom:10px; }
                .flt   { font-size:10px; color:#555; margin-bottom:10px; }
                table  { width:100%; border-collapse:collapse; }
                th     { background:#4472C4; color:#fff; padding:7px 10px; text-align:left; font-size:11px; }
                td     { padding:6px 10px; border-bottom:1px solid #eee; font-size:11px; }
                tr:nth-child(even) td { background:#f5f7ff; }
                .total-row td { background:#e8ecff!important; font-weight:bold; }
                @media print { @page { size:A4 landscape; margin:10mm; } }
            </style></head>
            <body>
                <h2>ATS SYNTHETIC (PVT.) LTD.</h2>
                <h3>Medical Summary Report</h3>
                <div class="meta">Printed: ${frappe.datetime.now_datetime()}</div>
                <div class="flt">${filters_html}</div>
                <table>
                    <thead>${thead}</thead>
                    <tbody>${tbody}<tr class="total-row">${total_row}</tr></tbody>
                </table>
            </body></html>`;

            var w = window.open("", "_blank");
            w.document.write(print_html);
            w.document.close();
            w.focus();
            setTimeout(() => w.print(), 500);
        });
    },

    formatter: function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        if (column.fieldname === "total_claimed" && data && data.total_claimed > 0) {
            value = `<span style="color:#5E64FF;font-weight:bold">${value}</span>`;
        }
        return value;
    },
};

function get_year_options() {
    var current = parseInt(frappe.datetime.get_today().split("-")[0]);
    var opts = [];
    for (var y = current; y >= current - 5; y--) opts.push(y);
    return opts.join("\n");
}